### PR TITLE
Implement torch.sort with cub::DeviceSegmentedRadixSort

### LINKED ISSUE #28871: Request to add stable sorting algorithm to PyTorch
## 🚀 Feature
My feature request is to add the algorithm of [stable sorting](https://en.wikipedia.org/wiki/Sorting_algorithm#Stability) to PyTorch. Stable sorting is a method of sorting an array that preserves order of elements with equal values. Formally, if there is a 1D tensor `A` and for some `i, j` `A[i] == A[j]`, then I would like to have `A[i]` earlier in the tensor than `A[j]` after its sorting. Generally speaking, the current implementation of `torch.sort` seems to be not preserving this order; for instance, see (produced by PyTorch 1.3.0):

```
In [2]: x = torch.Tensor([1, 2, 5, 4, 5])

In [3]: x.sort()
Out[3]: 
torch.return_types.sort(
values=tensor([1., 2., 4., 5., 5.]),
indices=tensor([0, 1, 3, 2, 4]))

In [4]: x.cuda().sort()
Out[4]: 
torch.return_types.sort(
values=tensor([1., 2., 4., 5., 5.], device='cuda:0'),
indices=tensor([0, 1, 3, 4, 2], device='cuda:0'))
```

## Motivation

This feature request is related to a problem I am facing outlined in https://discuss.pytorch.org/t/sorting-2d-tensor-by-pairs-not-columnwise/59465/8. Basically, I need to sort a 2D tensor of shape `(N, 2)` considering rows as tuples / pairs (row `A[i]` is less than row `A[j]`, iff `(A[i, 0] < A[j, 0]) or ((A[i, 0] == A[j, 0]) and (A[i, 1] <= A[j, 1]))`. There is a natural way to do this by sorting `A` by second column then stable sorting by first column. Please see https://discuss.pytorch.org/t/sorting-2d-tensor-by-pairs-not-columnwise/59465/8 for more detail.

If stable sort is implemented, it can be naturally extended to the sorting of tuples of any dimension via [Radix Sort](https://en.wikipedia.org/wiki/Radix_sort).

This issue seems to be related to a request about adding a stable sort to `torch.topk`: https://github.com/pytorch/pytorch/issues/27542 

## Pitch

Ideally, I would like to have a keyword argument in `torch.sort` or `torch.argsort`, for example `torch.sort(A, stable=True)`. There is already a similar syntax in NumPy: `np.sort(A, kind='stable')`. Stable sorting likely requires a different sorting algorithm than the one implemented in `torch.sort`.

## Alternatives

There is a simpler way to solve my problem suggested by @albanD but it does not suit my needs because of the possible precision loss (see https://discuss.pytorch.org/t/sorting-2d-tensor-by-pairs-not-columnwise/59465/8). If there is any other efficient solution to my problem that does not require stable sorting, please kindly post it here.

## Additional context

Here https://discuss.pytorch.org/t/is-torch-sort-stable/20714/3 is a small discussion about whether `torch.sort` is stable or not, and there is an answer 

> if the tensor contains at least 2049 elements, internally the function sortViaThrust is used, which as can be seen in the same file performs a stable sort.

I hope this can be useful.

cc @mruberry @rgommers

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Benchmark:
```python
import torch
import itertools

for i in range(1000):
    torch.arange(100000, device='cuda')

def run50_sync(f):
    for _ in range(50):
        f()
    torch.cuda.synchronize()

for i, j in itertools.product([512, 4096, 8192], repeat=2):
    print(i,j)
    t = torch.randn(i, j, device='cuda')
    torch.cuda.synchronize()
    %timeit run50_sync(lambda: torch.sort(t))
    torch.cuda.synchronize()
    %timeit run50_sync(lambda: torch.sort(t, dim=0))
    print()
```

Before
```
512 512
4.02 ms ± 28.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5 ms ± 15.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

512 4096
40.7 ms ± 74.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
33.9 ms ± 186 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

512 8192
71.7 ms ± 636 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
66.4 ms ± 163 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

4096 512
27.6 ms ± 27.8 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
46.6 ms ± 101 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

4096 4096
262 ms ± 1.14 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
321 ms ± 1.32 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

4096 8192
520 ms ± 5.47 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
661 ms ± 853 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

8192 512
54.6 ms ± 133 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
83.2 ms ± 320 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

8192 4096
521 ms ± 1.06 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
645 ms ± 1.47 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

8192 8192
1.04 s ± 2.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.34 s ± 541 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

After
```
512 512
4.65 ms ± 62.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
5.75 ms ± 62.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

512 4096
30.3 ms ± 261 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
39.4 ms ± 421 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

512 8192
59.7 ms ± 344 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
77 ms ± 601 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

4096 512
32.2 ms ± 376 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
37.1 ms ± 211 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

4096 4096
204 ms ± 471 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
272 ms ± 1.87 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

4096 8192
422 ms ± 3.25 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
562 ms ± 4.66 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

8192 512
63.1 ms ± 595 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
72.7 ms ± 532 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

8192 4096
401 ms ± 3.08 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
573 ms ± 2.59 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

8192 8192
831 ms ± 7.86 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.2 s ± 9.17 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

### CODE REVIEW

File: aten/src/ATen/cuda/cub.cuh [hunk 1]
```diff
@@ -53,7 +53,7 @@ struct cuda_type<c10::Half> {
 template<typename key_t>
 static inline void sort_keys(
     const key_t *keys_in, key_t *keys_out,
-    int64_t n, bool descending=false, int64_t start_bit=0, int64_t end_bit=sizeof(key_t)*8
+    int64_t n, bool descending=false, int64_t begin_bit=0, int64_t end_bit=sizeof(key_t)*8
```
Comment (zasdfgbnm) [inline aten/src/ATen/cuda/cub.cuh:56]: I renamed `start_bit -> begin_bit` to be consistent with cub

File: aten/src/ATen/cuda/cub.cuh [hunk 2]
```diff
@@ -65,35 +65,30 @@ static inline void sort_keys(
...
 
   auto allocator = c10::cuda::CUDACachingAllocator::get();
   c10::DataPtr keys_out_owner;
-  c10::DataPtr values_out_owner;
 
   if (keys_out == nullptr) {
     keys_out_owner = allocator->allocate(n * sizeof(key_t));
     keys_out = reinterpret_cast<key_t *>(keys_out_owner.get());
   }
-  if (values_out == nullptr) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/cuda/cub.cuh:93]: `values_out` will never be nullptr. If it is, then people should use `sort_keys` instead of `sort_pairs`

**Comment by ngimel** [comment]
We'd need a confirmation from cub that this is stable sort, because now we are using this path to support stable sort if requested by user.

**Review by ngimel — APPROVED** [review]
Thanks for the simplification!
